### PR TITLE
Fix SSLv3 Client Auth

### PR DESCRIPTION
**Pull request by mattcaswell**
The rec_layer_s3.c changes in this PR aren't really part of this change. They are being reviewed separately as a different standalone bug fix - but they are a pre-requisite for this fix, so they are included here too.

Client Auth for SSLv3 is badly broken. There are a number of fixes in this MR:

* Move the preparation of the client certificate to be post processing work after reading the CertificateRequest message rather than pre processing work prior to writing the Certificate message. As part of preparing the client certificate we may discover that we do not have one available. If we are also talking SSLv3 then we won't send the Certificate message at all. However, if we don't discover this until we are about to send the Certificate message it is too late and we send an empty one anyway. This is wrong for SSLv3.

* Having received a ClientKeyExchange message instead of a Certificate we know that we are not going to receive a CertificateVerify message. This means we can free up the handshake_buffer. However we better call ssl3_digest_cached_records() instead of just freeing it up, otherwise we later try and use it anyway and a core dump results. This could happen, for example, in SSLv3 where we send a CertificateRequest but the client sends no Certificate message at all. This is valid in SSLv3 (in TLS clients are required to send an empty Certificate message).

* In TLS if the server sends a CertificateRequest and the client does not provide one, if the server cannot continue it should send a HandshakeFailure alert. In SSLv3 the same should happen, but instead we were sending an UnexpectedMessage alert. This is incorrect - the message isn't unexpected - it is valid for the client not to send one - its just that we cannot continue without one.
The Client Auth tests were not correctly setting the Protocol, so that this aspect had no effect. It was testing the same thing lots of times for TLSv1.2 every time.

* In TLS during ClientAuth if the CA is not recognised you should get an UnknownCA alert. In SSLv3 this does not exist and you should get a BadCertificate alert. The tests didn't take account of this.
If configuring for anything other than the default TLS protocols then test failures were occurring.

The core dump issue noted above was found using the BoringSSL test suite, and is what started me on this odyssey. We'd have found it using our own client-auth tests if it wasn't for the fact that:

* We don't build sslv3 by default
* The tests were broken for sslv3 anyway
* Even if the obvious test problems were fixed it still would have been hidden because the tests weren't actually testing different protocol versions even though they appeared to be.

### CODE REVIEW

File: test/recipes/80-test_ssl_new.t [hunk 1]
```diff
@@ -39,6 +39,7 @@ my $no_dtls = alldisabled(available_protocols("dtls"));
 
 my %conf_dependent_tests = (
   "02-protocol-version.conf" => !$is_default_tls,
+  "04-client_auth.conf" => !$is_default_tls,
```
Comment (bryongloden) [inline test/recipes/80-test_ssl_new.t]: @mattcaswell. Nice

File: ssl/statem/statem_clnt.c [hunk 2]
```diff
@@ -175,7 +175,7 @@ int ossl_statem_client_read_transition(SSL *s, int mt)
             } else {
                 ske_expected = key_exchange_expected(s);
                 if (ske_expected < 0)
-                    return 0;
+                    goto err;
```
Comment (ekasper) [inline ssl/statem/statem_clnt.c]: Hm, key_exchange_expected (static method) documents it may return -1, but never does. Could you just make it a 1/0 boolean instead to simplify the logic?

File: ssl/statem/statem.c [hunk 3]
```diff
@@ -531,8 +531,7 @@ static SUB_STATE_RETURN read_state_machine(SSL *s) {
              * to that state if so
              */
             if(!transition(s, mt)) {
-                ssl3_send_alert(s, SSL3_AL_FATAL, SSL3_AD_UNEXPECTED_MESSAGE);
-                SSLerr(SSL_F_READ_STATE_MACHINE, SSL_R_UNEXPECTED_MESSAGE);
+                ossl_statem_set_error(s);
```
Comment (ekasper) [inline ssl/statem/statem.c]: I don't know the state machine code very well - Could you explain why it is always correct to remove this alert? What method is instead responsible for setting the alert if we don't get the expected handshake message?

Comment (ekasper) [inline ssl/statem/statem.c]: Ah, nvm - I now see that you moved this to _both_ the client and server.

File: test/ssl-tests/04-client_auth.conf.in [hunk 4]
```diff
@@ -26,15 +26,23 @@ sub generate_tests() {
     foreach (0..$#protocols) {
         my $protocol = $protocols[$_];
         my $protocol_name = $protocol || "flex";
+        my $caalert;
         if (!$is_disabled[$_]) {
+            if ($protocol_name eq "SSLv3") {
```
Comment (ekasper) [inline test/ssl-tests/04-client_auth.conf.in]: Hypothetically, "flex" would end up negotiating SSLv3 if that is the only enabled protocol.

However, I'm not entirely sure we should bother testing this contrived configuration. Up to you how correct you want to be there. Possibly more things break in that config anyway.

File: test/ssl-tests/04-client_auth.conf.in [hunk 5]
```diff
@@ -26,15 +26,23 @@ sub generate_tests() {
...
+                $caalert = "BadCertificate";
+            } else {
+                $caalert = "UnknownCA";
+            }
             # Sanity-check simple handshake.
             push @tests, {
                 name => "server-auth-${protocol_name}",
                 server => {
-                    "Protocol" => $protocol
```
Comment (ekasper) [inline test/ssl-tests/04-client_auth.conf.in]: Ooops! My bad.

File: ssl/statem/statem_srvr.c [hunk 6]
```diff
@@ -114,6 +114,17 @@ int ossl_statem_server_read_transition(SSL *s, int mt)
                 return 1;
             }
         }
+        if (mt == SSL3_MT_CLIENT_KEY_EXCHANGE && s->s3->tmp.cert_request
```
Comment (ekasper) [inline ssl/statem/statem_srvr.c]: It's really hard to parse this decision tree. Do you think you can reorganize it so all the `mt == SSL3_MT_CLIENT_KEY_EXCHANGE`logic falls under one `if`-clause. Something like:

```
if (mt == SSL3_MT_CLIENT_KEY_EXCHANGE) {
    if (s->s3->tmp.cert_request) {
        if (s->version == SSL3_VERSION) {
            /* look at verify mode */
        } else {
            /* fail: client should always send a Certificate message. */
        }
    } else {
        /* No certificate requested and none received: ok. */
    }
}
```

I think this separates the SSLv3 special casing a little better, what do you think?

Comment (ekasper) [inline ssl/statem/statem_srvr.c]: Better yet, maybe also separate all that hairy logic (also the `else if` branch) into  a separate `client_certificate_expected` method.

**Comment by ekasper** [comment]
Also enable ssl3 in one of the travis builds? I'd suggest one of the sanitizers.

File: ssl/statem/statem_clnt.c (hunk 2 above)
Comment (mattcaswell) [inline ssl/statem/statem_clnt.c]: Ahh - good point. Looks like it used to, but that logic was removed by 361a119127.

File: ssl/statem/statem_srvr.c (hunk 6 above)
Comment (mattcaswell) [inline ssl/statem/statem_srvr.c]: Yes, your `if` structure is better. However, I'm not convinced about the separate `client_certificate_expected` method idea. The problem is that, to actually make things simpler, you would also have to offload the "send a handshake failure alert instead of an unexpected message" processing to that function too. I'm not sure that is a good idea.

File: test/ssl-tests/04-client_auth.conf.in (hunk 4 above)
Comment (mattcaswell) [inline test/ssl-tests/04-client_auth.conf.in]: I think this is not worth addressing within the scope of this MR. Like you say I suspect other stuff breaks in that case...and I'm not sure if we care about it anyway.

**Comment by mattcaswell** [comment]
New commits pushed. I think I have addressed all comments:

* New commit added to remove the "-1" return value handling for `key_exchange_expected()`
* New commit added to simplify the server side CKE transition processing logic
* New commit added to switch on SSLv3 testing in Travis (lets see what happens with this!)

I also rebased to remove all the rec_layer_s3_.c changes which were showing up in this MR but were actually being reviewed elsewhere. Those have now been pushed to master so after the rebase, these no longer show up.

File: ssl/statem/statem_srvr.c [hunk 7]
```diff
@@ -101,13 +101,28 @@ int ossl_statem_server_read_transition(SSL *s, int mt)
...
-                && (!s->s3->tmp.cert_request
-                    || (!((s->verify_mode & SSL_VERIFY_PEER) &&
-                          (s->verify_mode & SSL_VERIFY_FAIL_IF_NO_PEER_CERT))
-                        && (s->version == SSL3_VERSION)))) {
-            st->hand_state = TLS_ST_SR_KEY_EXCH;
-            return 1;
+        if (mt == SSL3_MT_CLIENT_KEY_EXCHANGE) {
+            if (s->s3->tmp.cert_request) {
+                if (s->version == SSL3_VERSION) {
+                    if ((s->verify_mode & SSL_VERIFY_PEER) &&
```
Comment (ekasper) [inline ssl/statem/statem_srvr.c]: OOC, how can it happen that `SSL_VERIFY_FAIL_IF_NO_PEER_CERT` is set but `SSL_VERIFY_PEER` is not?

(Also, bad indent, and our style guide says to break before `&&` even though we all seem to prefer to break after...)

File: ssl/statem/statem_srvr.c [hunk 8]
```diff
@@ -101,13 +101,28 @@ int ossl_statem_server_read_transition(SSL *s, int mt)
...
+                          (s->verify_mode & SSL_VERIFY_FAIL_IF_NO_PEER_CERT)) {
+                        /*
+                         * This isn't an unexpected message as such - we're just
+                         * not going to accept it.
```
Comment (ekasper) [inline ssl/statem/statem_srvr.c]: Maybe add "... accept it because we require a client cert".

**Comment by ekasper** [comment]
+1 modulo the two minor  comments.

File: ssl/statem/statem_srvr.c (hunk 7 above)
Comment (mattcaswell) [inline ssl/statem/statem_srvr.c]: It makes no sense for SSL_VERIFY_FAIL_IF_NO_PEER_CERT to be set without SSL_VERIFY_PEER. The documentation says you must set both if you want SSL_VERIFY_FAIL_IF_NO_PEER_CERT. SSL_CTX_set_verify() doesn't enforce that though.

I couldn't see a bad indent? I changed to break before "&&"

File: ssl/statem/statem_srvr.c (hunk 8 above)
Comment (mattcaswell) [inline ssl/statem/statem_srvr.c]: Done

**Comment by mattcaswell** [comment]
New commit pushed. Can you just reconfirm your +1 because I wasn't sure about the bad indent you referred to.

**Comment by ekasper** [comment]
+1, whatever I thought I saw wrong with the indent, it's gone now :)

**Comment by mattcaswell** [comment]
Pushed. Thanks for the review.
